### ISSUE TITLE
After completing the training, how to evaluate the best checkpoint model for inference?

### ISSUE THREAD

**Issue by alancarlosml** [xuannianz/EfficientDet#217]
I would like to know how to choose the training model to make the inference? It always be the last one? Or the model with the greatest classification_loss or  val_greatest classification_loss?

Thanks!

**Comment by HMFazleRabbi** [comment]
First you have to select either inference.py or inference_quad.py as your inference script. Then change the model path in that file
`model_path = 'efficientdet-d1.h5'` 
change to 
`model_path = <your checkpoint file>'` 

The process is similar for evaluation. You just need to make the same change to eval/coco.py file

**Comment by alancarlosml** [comment]
@HMFazleRabbi thank you for the reply.

Yes, I did it. But my question is about which model of the following I should choose to replace here `model_path = <your checkpoint file>'`

![image](https://user-images.githubusercontent.com/413231/95613129-d56cf280-0a3a-11eb-8c19-108ddc3a4bb5.png)

**Comment by HMFazleRabbi** [comment]
Usually, we choose the one with the lowest validation loss which would be _coco_50_0.0285_0.9034.h5_. It has a training loss of 0.0285 and validation loss of 0.9034, however there is a chance of over fitting. So, my advise would be to check the tensorboard graphs for indication of over fitting.

**Comment by alancarlosml** [comment]
OK, @HMFazleRabbi.

One more help, please. Checking my tensorboard graphs I got this:

![image](https://user-images.githubusercontent.com/413231/96058649-2e38f280-0e62-11eb-985a-db5e1bf15a21.png)
![image](https://user-images.githubusercontent.com/413231/96058669-3e50d200-0e62-11eb-9c00-7276622ce0af.png)
![image](https://user-images.githubusercontent.com/413231/96058738-6b04e980-0e62-11eb-9716-9facf56894b3.png)

My validation graphs are very irregular. Is this can be considered overfitting?

Thank you in advance.

**Comment by HMFazleRabbi** [comment]
What was the command that you used to run the training?

**Comment by alancarlosml** [comment]
I'm training with my own data (COCO format) the following command:

`python train.py --snapshot imagenet --phi 3 --gpu 0 --random-transform --compute-val-loss --freeze-backbone --batch-size 8 --steps 1377 coco breast_cancer_dir`

**Comment by eypros** [comment]
To make a general comment as an abstract approach the lowest val loss seem a valid candidate to test BUT it's not necessary the one that actually would perform the best taking into account the mAP (mean Average Precision). 
The implementation provided in the code does not perform mAP calculation in Coco mode as in Pascal (it performs an mAP evaluation using the provided eval set at the end of each epoch on the expense of slower training of course). This way you basically can get the best performance on the fly.

But another comment is that in your case, with the custom data you used, the difference between training and validation loss is colossal. This could mean overfitting for example. 0.0285 vs 0.90! It could also mean too few training samples and so the model learns them very well and do not generalize well enough. Or a more extreme data augmentation could be tested also.
